crates/bevy_utils/src/label.rs

```diff
@@ -60,90 +66,40 @@ where
-                std::any::TypeId::of::<Self>()
-            }
-
-            /// Clones this `
-            #[doc = stringify!($label_trait_name)]
-            /// `
-            fn dyn_clone(&self) -> Box<dyn $label_trait_name>;
-
-            /// Casts this value to a form where it can be compared with other type-erased values.
-            fn as_dyn_eq(&self) -> &dyn ::bevy_utils::label::DynEq;
-
-            /// Feeds this value into the given [`Hasher`].
-            ///
-            /// [`Hasher`]: std::hash::Hasher
-            fn dyn_hash(&self, state: &mut dyn ::std::hash::Hasher);
-        }
+/// This is expected to have sufficient entropy to be different for each value.
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct UniqueValue(TypeId, u64);
 
-        impl PartialEq for dyn $label_trait_name {
-            fn eq(&self, other: &Self) -> bool {
-                self.as_dyn_eq().dyn_eq(other.as_dyn_eq())
-            }
-        }
-
-        impl Eq for dyn $label_trait_name {}
-
-        impl ::std::hash::Hash for dyn $label_trait_name {
-            fn hash<H: ::std::hash::Hasher>(&self, state: &mut H) {
-                self.dyn_hash(state);
-            }
-        }
-
-        impl ::std::convert::AsRef<dyn $label_trait_name> for dyn $label_trait_name {
-            #[inline]
-            fn as_ref(&self) -> &Self {
-                self
-            }
-        }
-
-        impl Clone for Box<dyn $label_trait_name> {
-            fn clone(&self) -> Self {
-                self.dyn_clone()
-            }
-        }
-
-        impl $label_trait_name for Box<dyn $label_trait_name> {
-            fn inner_type_id(&self) -> ::std::any::TypeId {
-                (**self).inner_type_id()
-            }
+impl UniqueValue {
+    fn of<T: DynHash + ?Sized>(value: &T) -> Self {
+        let mut hasher = DefaultHasher::new();
+        value.dyn_hash(&mut hasher);
+        let hash = hasher.finish();
 
-            fn dyn_clone(&self) -> Box<dyn $label_trait_name> {
-                // Be explicit that we want to use the inner value
-                // to avoid infinite recursion.
-                (**self).dyn_clone()
-            }
-
-            fn as_dyn_eq(&self) -> &dyn ::bevy_utils::label::DynEq {
-                (**self).as_dyn_eq()
-            }
+        Self(TypeId::of::<T>(), hash)
+    }
+}
 
-            fn dyn_hash(&self, state: &mut dyn ::std::hash::Hasher) {
-                (**self).dyn_hash(state);
-            }
-        }
-    };
+/// Returns a reference to the interned `Debug` string of `value`.
+///
+/// If the string has not been interned, it will be allocated in a [`Box`] and leaked, but
+/// subsequent calls with the same `value` will return the same reference.
+pub fn intern_debug_string<T>(value: &T) -> &'static str
+where
+    T: Debug + DynHash + ?Sized,
+{
+    let key = UniqueValue::of(value);
+    let mut map = INTERNER.get_or_init(default).write().unwrap();
```

Comment:
I'd like this to have a more descriptive error message instead of just using `unwrap`.

Reply:
Any suggestions? Is the standard message about the `RwLock` being poisoned not descriptive enough?

Reply:
IIRC, `.unwrap()` uses the `Debug` implementation of the error type, so the error message will just be `PoisonError { ... }`.

Maybe it would be better to bypass poisoning entirely by using `.unwrap_or_else(PoisonError::into_inner)`? If a panic occurs while the lock is being held, I believe it would still leave the interner in a valid state.